server/src/routes/transaction_controller.rs

```diff
@@ -0,0 +1,237 @@
+use actix_web::{web, HttpResponse, error::{self}, http::{StatusCode}, get, post};
+use diesel::{SqliteConnection, RunQueryDsl, r2d2};
+use log::debug;
+use serde::{Deserialize};
+use derive_more::{Display, Error};
+use diesel::prelude::*;
+
+use crate::{models::{Transaction, TransactionKind, User}};
+use crate::models::APIError;
+
+// TODO: move to database module
+type DbPool = r2d2::Pool<r2d2::ConnectionManager<SqliteConnection>>;
+
+#[derive(Deserialize)]
+pub struct SoloTransaction {
+	pub mutation: f32,

```

Comment:
At the API level should we enforce that all money is measured in cents and make the conversion the clients job, would simplify the server end parsing and avoid any potential issues with floating-point numbers (although I'm sure Rust does the right thing when converting FP to Int without rounding it would make every operation we're doing in the code explictly in cents). 

At the same time, since we have enums for the transaction type, should we limit it to unsigned integers? Then we never get into a  'adding a negative mutation' case.

Reply:
We talked about this, I thought we agreed on an absolutely not using cents. It'd just be confusing for the end user and makes using the API less easy, even if only marginally.

This way we're sure the decimals are handled the same way every time. Parsing to i32 does just straight up delete everything behind the comma btw. Still have to make that reflect in the return values, same for the get user call.

As an aside btw, we really shouldn't set the balance when creating a user, that's imo kinda weird. Just follow it up with a deposit.

The positive/negative decision here is basically made on the basis of how it reflects on the users balance. A deposit is positive cause it adds to the balance, a transfer negative cause it takes from the balance. Else it might get kind of tricky to remember imo.

Reply:
> We talked about this, I thought we agreed on an absolutely not using cents. It'd just be confusing for the end user and makes using the API less easy, even if only marginally.
> 

We did talk about, but I walked away with the exact opposite impression - that we should have all the moving parts communicate in cents and format _only_ for display to the user. The end user shouldn't care at all about what the client & server communicate the balance/transaction in. It's also written in the API Spec and the `User` API uses cents to communicate.

Using floats we are all but guaranteed to enter a situation in the future where we run into a CWE-1339 like situation (however small).

> This way we're sure the decimals are handled the same way every time. Parsing to i32 does just straight up delete everything behind the comma btw. Still have to make that reflect in the return values, same for the get user call.
> 

For sure the way it is currently, but it eliminates an entire potential bug class by using cents - floating point arithmatic cannot accurately represent what we're trying to handle and it provides no real value to use it over cents.

> As an aside btw, we really shouldn't set the balance when creating a user, that's imo kinda weird. Just follow it up with a deposit.
> 

That I agree with, I can make a new MR with that change (or we can include it here - whatever you prefer).

> The positive/negative decision here is basically made on the basis of how it reflects on the users balance. A deposit is positive cause it adds to the balance, a transfer negative cause it takes from the balance. Else it might get kind of tricky to remember imo.

That's fair - but I think we can simplify the internal API and potentially remove a bug class by not worrying about signs - a deposit it _always_ an addition of a positive amount (i.e. you can't deposit a negative amount), and a transaction _always_ spends a positive amount.

We already have the enum to determine what type of transaction is occuring - so we don't need to remember much beyond switching on the type.

Reply:
You know what I think I actually mostly agree with you. If we have the api users also use cents everywhere there's no floating point issues at all. 

I also agree on just using positive numbers, I think that works fine, and I also think it's a good idea to move to unsigned integers for the api since we can then communicate that in an easier way. With the caveat that we have to actively check for the max value since the max value of u64 does not match i64 and sqlite doesn't let us save unsigned integers in the database. (Mind you, I do want to save them the way we currently do in the database, including using negative numbers in the transactions to reflect balance changes so you can verify a user's balance with a simple query)

Reply:
> I also agree on just using positive numbers, I think that works fine, and I also think it's a good idea to move to unsigned integers for the api since we can then communicate that in an easier way. With the caveat that we have to actively check for the max value since the max value of u64 does not match i64 and sqlite doesn't let us save unsigned integers in the database. (Mind you, I do want to save them the way we currently do in the database, including using negative numbers in the transactions to reflect balance changes so you can verify a user's balance with a simple query)

Trying to think though ways to enforce the check of max values etc, maybe a custom Diesel type where we can check the values being passed into the DB?

Not opposed to storing the signed value in the DB - we can easily convert those once we know the transaction type.